### PR TITLE
Grating equation

### PR DESCRIPTION
**Pull request by nden**
### Description

This pull request implements the grating equation and solves it for the wavelength.
In this form the implementation is valid for small and oblique angles. When used with small angles
sine of the angles should be passed in as inputs. When used with oblique angles the inputs are the direction cosines.

I'd like to get an opinion on this by other potential users or others with spectroscopy experience.
@jehturner @stscicrawford 

Edit: The best reference I can find for this implementation is 
"Description of Diffraction Grating Behavior in Direction Cosine Space" by James E. Harvey and Cynthia L. Vernold in  Applied Optics, Vol. 73, Issue 9, pp. 1105-1112 (1983), linked below, though I'm not sure it's visible by everyone. I need to use it in cosine space.

https://www.osapublishing.org/ao/fulltext.cfm?uri=ao-37-34-8158&id=170236

### CODE REVIEW

File: astropy/modeling/physical_models.py [hunk 1]
```diff
@@ -194,3 +194,67 @@ def lambda_max(self):
     def nu_max(self):
         """Peak frequency when the curve is expressed as power density."""
         return 2.8214391 * const.k_B * self.temperature / const.h
+
+
+class WavelengthFromGratingEquation(Model):
+    """
+    Solve the Grating Dispersion Law for the wavelength.
+
+    Note: This form of the equation can be used for paraxial
```
Comment (pllim) [inline astropy/modeling/physical_models.py]: Please use the `.. note::` Sphinx directive.

File: astropy/modeling/physical_models.py [hunk 2]
```diff
@@ -194,3 +194,67 @@ def lambda_max(self):
...
+    (small angle approximation) as well as oblique incident angles.
+    With paraxial systems the inputs are sin of the angles and it
+    transforms to :math: \sin(alpha_in) + \sin(alpha_out) / m * d .
```
Comment (pllim) [inline astropy/modeling/physical_models.py]:
```suggestion
    transforms to :math:`\sin(alpha_in) + \sin(alpha_out) / m * d`.
```

File: astropy/modeling/physical_models.py [hunk 3]
```diff
@@ -194,3 +194,67 @@ def lambda_max(self):
     def nu_max(self):
         """Peak frequency when the curve is expressed as power density."""
         return 2.8214391 * const.k_B * self.temperature / const.h
+
+
+class WavelengthFromGratingEquation(Model):
+    """
```
Comment (pllim) [inline astropy/modeling/physical_models.py]:
```suggestion
    r"""
```

**Comment by nden** [comment]
Thanks @pllim , will do. This is why it's still WIP. I'd like to get feedback on the science before finishing it.

File: astropy/modeling/physical_models.py [hunk 4]
```diff
@@ -194,3 +194,67 @@ def lambda_max(self):
...
+    def __init__(self, groove_density, spectral_order, **kwargs):
+        super().__init__(groove_density=groove_density,
+                         spectral_order=spectral_order, **kwargs)
+        self.inputs = ("alpha_in", "alpha_out")
+        """ Sine function of the angles or the direction cosines."""
+        self.outputs = ("wavelength",)
+        """ Wavelength."""
+
+    def evaluate(self, alpha_in, alpha_out, groove_density, spectral_order):
+        result = alpha_in + alpha_out / (groove_density.value * spectral_order)
```
Comment (nden) [inline astropy/modeling/physical_models.py]: I missed the parenthesis in the nominator. That said I'd very much like to hear about this approach from spectroscopists.

**Comment by bsipocz** [comment]
@jradavenport - this might interest you to review

**Comment by crawfordsm** [comment]
@nden This looks like the right track to me.  I like the clear example with showing how it can be used and involved.  

I was looking at how I used this in some other code, and it still allows for chaining together with other factors if needed, but would be able to use this fairly straightforwardly in the existing code.   

Thanks for submitting the PR

**Comment by nden** [comment]
@bsipocz I will move these models to `gwcs`. They are probably going to be used only in the context of WCS anyway.

**Comment by jehturner** [comment]
Sorry for the slow reply, while we were making the first release of our Python pipeline for general use, not to mention having a week or two of national rioting and a 4-day weekend. Does this still need input?

I don't really have experience with basing WCS on such a model, nor access to the linked article (though I could probably get hold of it if needed), but it does seem useful to have a standard grating model. Our old IRAF code ended up with undocumented, ad hoc wavelength calculations here and there that are sometimes hard to follow... But what about the rest of the calculation needed to get to wavelength from some practical unit like pixels? Can that be usefully generalized?

Our calculations in IRAF tend to be used for purposes like determining the limits of a spectrum or estimating its central wavelength and dispersion prior to fitting an empirical wavelength model. Perhaps those things should be re-cast as (approximate?) WCS operations, consistent with what you say. It also seems plausible, off the top of my head, that combining the grating equation with an empirical distortion model could help fit wavelength solutions more robustly (our Python prototype is already half-way there, with a distortion model split from the wavelength mapping itself).

Could the doc string be more explicit about the limits within which the sine vs direction cosines should be used? Also, this code requires separate `alpha_in` & `alpha_out` inputs but then merely adds them together to get the numerator of the equation. Sometimes it seems convenient to provide an equivalent form of that numerator, `2 cos(K) sin(φ'-tilt)`, where `K` & `φ'` are fixed properties/conventions of the spectrograph. That expression may be partly pre-calculated or come entirely from a look-up table, in which case you can't derive `alpha_in` and `alpha_out` to give to this model separately, but you **could** pass `alpha_in + alpha_out` because it's just the same thing you have calculated using the other form. Of course, this would leave you with rather a trivial model that doesn't actually do much of the work... I suppose the alternative is to have multiple models for different cases. Anyway, just some thoughts...

It would seem instructive to compare with S5 of FITS WCS paper III. The grism equation used there is more general (but I'd need to remind myself of the details).
